src/docs/commands_keybindings.md

```diff
@@ -1,122 +1,141 @@
+
+## Contributing commands
+
+By contributing a command, you can add a custom action to Theia. A command can be triggered by the user via the quick access bar, a keybinding or a menu entry. It can also be called programmatically.
+
+All commands of a Theia application are managed in the `CommandRegistry`. To contribute commands to the command registry, modules must implement the `CommandContribution` interface (see code example below). 
+
+A command is an object with an `id` and an optional, potentially user visible `label` (see HelloworldCommand in the example below). The command contribution receives the `CommandRegistry` in the `registerCommands` function as a parameter. The command can then be registered by calling `registerCommand` on this registry. Along with the command a callback needs to be provided which is executed whenever the command is triggered (`CommandHandler`). In the example, the command uses the MessageService to “say hello”. 
+
+**helloworld-contribution.ts**
 ```typescript
+export const HelloworldCommand: Command = {
+   id: 'Helloworld.command',
+   label: "Say Hello"
+};
+ 
 @injectable()
-export class JavaCommandContribution implements CommandContribution {
-...
-    registerCommands(commands: CommandRegistry): void {
-        commands.registerCommand(SHOW_JAVA_REFERENCES, {
-            execute: (uri: string, position: Position, locations: Location[]) =>
-                commands.executeCommand(SHOW_REFERENCES.id, uri, position, locations)
-        });
-        commands.registerCommand(APPLY_WORKSPACE_EDIT, {
-            execute: (changes: WorkspaceEdit) =>
-                !!this.workspace.applyEdit && this.workspace.applyEdit(changes)
-        });
-    }
+export class HelloworldCommandContribution implements CommandContribution {
+ 
+   constructor(
+       @inject(MessageService) private readonly messageService: MessageService,
+   ) { }
+ 
+   registerCommands(registry: CommandRegistry): void {
+       registry.registerCommand(HelloworldCommand, {
+           execute: () => this.messageService.info('Hello World!')
+       });
+   }
 }
+
 ```
 
-Each contribution needs a unique command id and a command handler (the callback that will be executed).
+To make the command execution context sensitive, the `CommandHandler` can optionally implement `isEnabled` and `isVisible`. When a command is executed, the command registry check all registered handlers. The first handler that returns true on `isEnabled` will be considered to be active and it will be executed. Only one handler should be active (`isEnabled === true`) at the same time, though. `isVisible` controls the visibility of menu items that are connected to a command, if the active handler returns true, the menu item will be visible and vice versa.
```

Comment:
```suggestion
To make the command execution context sensitive, the `CommandHandler` can optionally implement `isEnabled` and `isVisible`. When a command is executed, the command registry checks all registered handlers. The first handler that returns true on `isEnabled` will be considered to be active and it will be executed. Only one handler should be active (`isEnabled === true`) at the same time, though. `isVisible` controls the visibility of menu items that are connected to a command, if the active handler returns true, the menu item will be visible and vice versa.
```